app/services/stripe_model_callbacks/invoice/updated_service.rb

```diff
@@ -1,23 +1,39 @@
 class StripeModelCallbacks::Invoice::UpdatedService < StripeModelCallbacks::BaseEventService
-  def execute
-    invoice.assign_from_stripe(object)
+  TRACKED_ACTIVITIES = {
+    "invoice.payment_failed": :payment_failed,
+    "invoice.payment_succeeded": :payment_succeeded,
+    "invoice.sent": :sent,
+    "invoice.upcoming": :upcoming
+  }.freeze.with_indifferent_access
+  private_constant :TRACKED_ACTIVITIES
 
-    if invoice.save
-      create_activity
+  def execute
+    # The difference between the stripe events is about a few milliseconds - with advisory_lock
+    # we will prevent from creating invoice duplicates due to race condition.
+    # https://stripe.com/docs/webhooks/best-practices#event-ordering
+    StripeModelCallbacks::ApplicationRecord.with_advisory_lock("stripe_invoice-id#{object.id}") do
```

Comment:
This is an issue with all types of records, right? Then it should probably go inside this file as a helper method:
https://github.com/kaspernj/stripe_model_callbacks/blob/master/app/services/stripe_model_callbacks/base_service.rb

Maybe name it `with_object(object)`?

Then all services that handels Stripe objects should call it, to make sure two objects aren't handled at the same time, which should effectively shut down this problem across all records for good?

Reply:
@kaspernj It looks like it is. I will create a lock_name base on object.object and object.id. But in some cases it won't work, because for example `customer.discount.created` don't have id, but the coupon inside discount object does. Maybe you know any other examples. It would be helpful to implement a proper solution.

Reply:
One more thing, we can do it like this:
```
  def self.execute_with_advisory_lock!(*args, &blk)
    with_exception_notifications do
      StripeModelCallbacks::ApplicationRecord.with_advisory_lock(advisory_lock_name(*args)) do
        response = execute(*args, &blk)
        raise response.errors.join(". ") unless response.success?

        response
      end
    end
  end
```
WDYT?

Reply:
In the case where the object doesn't have an ID, maybe we should name a lock for the entire model class for that object? So any discount couldn't be handled at the same time?

Reply:
Otherwise it looks good 👍

Reply:
@kaspernj please take a look on pushed changes :)